tfjs-backend-webgpu/src/fft_webgpu.ts

```diff
@@ -0,0 +1,82 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {getMainHeaderString as main, WebGPUProgram} from './webgpu_program';
+import {computeDispatch, flatDispatchLayout} from './webgpu_util';
+
+export class FFTProgram implements WebGPUProgram {
+  variableNames: string[] = ['real', 'imag'];
+  outputShape: number[] = [];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  uniforms = 'exponentMultiplier : f32, denominator: f32,';
+  workgroupSize: [number, number, number] = [64, 1, 1];
+  size = true;
+  component: string;
+
+  constructor(component: 'real'|'imag', shape: [number, number]) {
+    this.outputShape = shape;
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workgroupSize);
+
+    this.component = component;
+    this.shaderKey = `fft_${component}`;
+  }
+
+  getUserCode(): string {
+    const opString = this.component === 'real' ?
+        'return real * expR - imag * expI;' :
+        'return real * expI + imag * expR;';
+    const userCode = `
+    fn unaryOpComplex(real: f32, expR: f32, imag: f32, expI: f32) -> f32 {
+      ${opString}
+    }
+
+    fn mulMatDFT(batch: i32, index: i32) -> f32 {
+      let indexRatio = f32(index) / f32(uniforms.realShape[1]);
+      let exponentMultiplierTimesIndexRatio =
+          uniforms.exponentMultiplier * indexRatio;
+
+      var result = 0.0;
+
+      for (var i = 0; i < uniforms.realShape[1]; i = i + 1) {
+        // x = (-2|2 * PI / N) * index * i;
+        let x = exponentMultiplierTimesIndexRatio * f32(i);
+        let expR = cos(x);
+        let expI = sin(x);
+        let real = getReal(batch, i);
+        let imag = getImag(batch, i);
```

Comment:
Is this a direct mapping from the definition of discrete Fourier transform:
$$X_k = \sum_{n=0}^{N-1} x_n e^{-\frac{2kn\pi{}i}{N}}$$

It has $\mathcal{O}(N^2)$ complexity, so the first F in FFT ( $\mathcal{O}(N\log{}N)$ ) is lost. FFT should be super suitable for parallelism using the butterfly diagram, and could be a good example where WebGPU beats WebGL, though I have never remembered the details of the algorithm :sweat:

Reply:
Yes, It directly uses DFT to implement FFT here. FFT implementation is more difficult, I plan to implementation it in the future as an optimization task.